llvm/lib/CodeGen/Yk/FixStackmapsSpillReloads.cpp

```diff
@@ -0,0 +1,211 @@
+// values in $rdi and $rsi into another register or onto the stack before the
+// call. Then immediately after the call it inserted instructions to reload
+// the spilled values back into the original registers. Since during
+// deoptimisation we return to immediately after the call, the stackmap is now
+// tracking the wrong values, e.g. in this case $rdi and $rsi instead of the
+// spill locations.
+//
+// This pass interates over all basic blocks, finds spill reloads inserted
+// inbetween a call and stackmap, replaces the stackmap operands with the
+// spill reloads, and then moves the stackmap instruction up just below the
+// call.
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CodeGen/MachineBasicBlock.h"
+#include "llvm/CodeGen/MachineFrameInfo.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/StackMaps.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/IR/DebugLoc.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "fix-stackmaps-spill-reloads"
+
+namespace {
+
+class FixStackmapsSpillReloads : public MachineFunctionPass {
+public:
+  static char ID;
+
+  FixStackmapsSpillReloads() : MachineFunctionPass(ID) {
+    initializeFixStackmapsSpillReloadsPass(*PassRegistry::getPassRegistry());
+  }
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  StringRef getPassName() const override {
+    return "Stackmaps Fix Post RegAlloc Pass";
+  }
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+};
+
+} // namespace
+
+char FixStackmapsSpillReloads::ID = 0;
+char &llvm::FixStackmapsSpillReloadsID = FixStackmapsSpillReloads::ID;
+
+INITIALIZE_PASS_BEGIN(FixStackmapsSpillReloads, DEBUG_TYPE, "Fixup Stackmap Spills",
+                      false, false)
+INITIALIZE_PASS_END(FixStackmapsSpillReloads, DEBUG_TYPE, "Fixup Stackmap Spills",
+                    false, false)
+
+
+bool FixStackmapsSpillReloads::runOnMachineFunction(MachineFunction &MF) {
+  bool Changed = false;
+  const TargetInstrInfo *TII = MF.getSubtarget().getInstrInfo();
+  for (MachineBasicBlock &MBB : MF) {
+    bool Collect = false;
+    std::set<MachineInstr *> Erased;
+    MachineInstr *LastCall = nullptr;
+    std::map<Register, MachineInstr *> Spills;
+    for (MachineInstr &MI : MBB) {
+      if (MI.isCall() && !MI.isInlineAsm()) {
+        // YKFIXME: Do we need to check for intrinsics here or have they been
+        // removed during lowering?
+        if (MI.getOpcode() != TargetOpcode::STACKMAP &&
+            MI.getOpcode() != TargetOpcode::PATCHPOINT) {
+          Collect = true;
```

Comment:
Can we (in general) have a few more comments around, so that a casual reader knows what each step does at a glance?

Reply:
e.g. What is `Collect` and `Erase`? Why are we setting `Collect` here?

Reply:
See if this helps: [21e6574](https://github.com/ykjit/ykllvm/pull/61/commits/21e65743b33587cb85486cdab029d38b6e62fd87)